### PR TITLE
Introduce docs/development/style_guide.md

### PR DESCRIPTION
**Pull request by ScottTodd**
This adds a style guide with some recommendations for authoring CMake, Python, Bash, and GitHub Actions (.yml) files.

I started with some common patterns that have come up in recent code reviews then got some help from https://www.claude.com/product/claude-code filling in details and concrete positive/negative examples (Claude got the general structure from reading code across the project and then I made adjustments).

Looking for feedback on what to include here and how strong to make the language. My main goals are:

1. Having a self-service reference page with example code that contributors (or agents) can reference while making changes, not just "follow the rest of the project's style"
2. Having somewhere we can link to during code reviews where there are disagreements

### CODE REVIEW

File: docs/development/style_guide.md [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+directory.
+
+Benefits:
+
+- **Location-independent:** Script works from any directory
+- **Explicit:** Clear where files are relative to the script
+- **CI-friendly:** Works in CI environments with varying working directories,
+  especially when scripts and workflows are used in other repositories
+
+❌ **Bad:** Assuming script runs from a specific directory
```
Comment (marbre) [inline docs/development/style_guide.md]:
```suggestion
❌ **No:** Assuming script runs from a specific directory
```
Maybe flip the order and first give the positive sample before showing what not to do?

File: docs/development/style_guide.md [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+from pathlib import Path
+
+# Assumes script is run from repository root
+config_file = Path("build_tools/config.json")
+
+# Assumes script is run from its own directory
+data_file = Path("../data/artifacts.tar.gz")
+```
+
+✅ **Good:** Paths relative to the script location
```
Comment (marbre) [inline docs/development/style_guide.md]:
```suggestion
✅ **Yes:** Paths relative to the script location
```

File: docs/development/style_guide.md [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+)
+```
+
+## Bash guidelines
+
+> [!WARNING]
+> Bash is **strongly discouraged** for nontrivial usage in .yml GitHub Actions
+> workflow files and script files.
+>
+> Prefer to use Python scripts in most cases.
```
Comment (marbre) [inline docs/development/style_guide.md]:
```suggestion
> Use Python scripts in most cases instead.
```

**Review by marbre — COMMENTED** [review]
Nice! I really appreciate this. The main suggestions I have would be to rename Bad and Good to Yes and No, respectively and maybe start with the sample on what to do instead of showing what not to do first.

File: docs/development/style_guide.md (hunk 2 above)
Comment (ScottTodd) [reply docs/development/style_guide.md]: Scanning through https://github.com/Kristories/awesome-guidelines for style guide style tips (soooo meta)...

* https://google.github.io/styleguide/cppguide.html uses all green text for "good" and all red text for "bad", with no other labels
* https://google.github.io/styleguide/pyguide.html uses "yes" and "no", starting with "yes"
* https://peps.python.org/pep-0008/ uses "correct" and "wrong", starting with "correct"
* https://llvm.org/docs/CodingStandards.html uses "avoid" and "preferred/OK", starting with "avoid"
* https://firefox-source-docs.mozilla.org/code-quality/coding-style/coding_style_cpp.html has no standard verbiage but starts with negative examples. "don't do this:" "instead do this", "incorrect", "correct"
* https://alibaba.github.io/Alibaba-Java-Coding-Guidelines/ uses "Positive example" and "Counter example"

**How about "Preferred" and "Avoid", starting with "Preferred"?**

File: docs/development/style_guide.md [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+  - Software versions (e.g. Python)
+- Collaborate with upstream projects
+
+### Formatting using pre-commit hooks
+
+We enforce formatting for certain languages using
+[_pre-commit_](https://pre-commit.com/) with hooks defined in
+[`.pre-commit-config.yaml`](/.pre-commit-config.yaml). See the pre-commit checks
+[section in `CONTRIBUTING.md`](/CONTRIBUTING.md#pre-commit-checks) for
+instructions on setting up pre-commit.
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: i would add here also the instruction how to set it up. it is just the `pip install pre-commit` and `pre-commit install`. For more details and explanation they can look in the `CONTRIBUTING.md`

so there is no excuse of not using it.

File: docs/development/style_guide.md [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+([rocm-systems](https://github.com/ROCm/rocm-systems),
+[rocm-libraries](https://github.com/ROCm/rocm-libraries)), subprojects **must**
+be compatible with one another at the same git commit, and TheRock enforces
+this.
+
+## Python guidelines
+
+We generally follow the [PEP 8 style guide](https://peps.python.org/pep-0008/)
+using the [_Black_ formatter](https://github.com/psf/black). The guidelines here
+extend PEP 8 for our projects.
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:82]: should we add something like "pre-commit hook takes care of it to follow it"?

File: docs/development/style_guide.md [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+# Clear, readable, platform-independent
+artifact_path = Path(output_dir) / artifact_group / "rocm.tar.gz"
+
+# Concise and type-safe
+artifacts_dir = Path(base_dir) / "build" / "artifacts"
+if artifacts_dir.exists():
+    files = list(artifacts_dir.iterdir())
+```
+
+Benefits:
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: put the benefits above the code example and really highlight the platform independences - as that is the main issue

File: docs/development/style_guide.md [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+
+[Python pathlib documentation](https://docs.python.org/3/library/pathlib.html)
+
+❌ **Bad:** String-based path manipulation
+
+```python
+import os
+
+# Hard to read, platform-specific separators
+artifact_path = output_dir + "/" + artifact_group + "/" + "rocm.tar.gz"
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: artifact group not defined.. might not matter for this example but would be nice to see if that is a string or not?

File: docs/development/style_guide.md (hunk 1 above)
Comment (HereThereBeDragons) [reply docs/development/style_guide.md]: we should decide on good/bad vs yes/no? at least it should be consistent
and i agree: good stuff first before bad stuff

File: docs/development/style_guide.md [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+Use [`argparse`](https://docs.python.org/3/library/argparse.html) for
+command-line argument parsing with clear help text and type conversion.
+
+Benefits:
+
+- **Automatic help:** Users get `-h/--help` for free
+- **Type conversion:** Arguments are converted to correct types
+- **Validation:** Required arguments are enforced
+
+❌ **Bad:** Manual argument parsing
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: just remove the entire manual argument parsing section. just show how it is done in argparse.

for other sections the good/bad part are helpful but here i think this could only give wrong ideas. unless you come from C you dont do manual parsing anyway

File: docs/development/style_guide.md [hunk 9]
```diff
@@ -0,0 +1,583 @@
...
+```
+
+✅ **Good:** Using `argparse` with types and defaults
+
+```python
+import argparse
+from pathlib import Path
+
+
+def fetch_artifacts(output_dir: Path):
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: remove this function

File: docs/development/style_guide.md [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(description="Fetches artifacts")
+    parser.add_argument(
+        "--output-dir",
+        type=Path,
+        default=Path("build/artifacts"),
+        help="Output path for fetched artifacts (default: build/artifacts)",
+    )
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:198]: add some parsing for booleans? therea different ways. but preferred should be the `BooleanOptionalAction`

e.g.
```
parser.add_argument(
        "--include-tests",
        default=False,
        required=False,
        action=argparse.BooleanOptionalAction,
        help="""Invert behavior: create a list of tests to include (run) instead of skip.
Output can be used with 'pytest -k <list>'""",
    )
````

File: docs/development/style_guide.md [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+from pathlib import Path
+
+
+def fetch_artifacts(run_id: int, output_dir: Path) -> list[Path]:
+    """Fetch artifacts from the given run ID."""
+    # Implementation here
+    return artifacts
+
+
+def main(argv: list[str]) -> int:
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:327]: i find this a delicate topic. especially looking at the current state of the scripts.

we should really highlight here:
- write the script in a way that you can run it from cli and directly call it from another script
- have meaningfully named entry functions
- dont take argparse arguments directly as input variables for your entryfunction
( - I like to have argparse in a separate function but i guess `main()` is also fine)

what i do not like is what e.g. `fetch_artifacts.py` is currently doing. 

```py
def run(args):
    run_github_repo = args.run_github_repo
    run_id = args.run_id
    artifact_group = args.artifact_group
    output_dir = args.output_dir
    output_dir.mkdir(parents=True, exist_ok=True)

    <do stuff>

def main(argv):
    parser = argparse.ArgumentParser(prog="fetch_artifacts")
    <parse arguments>

    run(args)


if __name__ == "__main__":
    main(sys.argv[1:])
```

i would strongly encourage here that future scripts would not accept `def run(args)` but would insist on `def run(run_github_repo, run_id, artifact_group, output_dir)` with type hinting. and maybe also rename `run`  as the entry function to something more meaningful if reasonable.

e.g. without explicit `main()`

```py
def get_tests(
    amdgpu_family: str = "", pytorch_version: str = "", create_skip_list: bool = True
) -> str:
<some implementation>

if __name__ == "__main__":
    args = cmd_arguments(sys.argv[1:])

    tests = get_tests(args.amdgpu_family, args.pytorch_version, not args.include_tests)
````

File: docs/development/style_guide.md [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+    overwrite=True,
+    validate=False,
+    compress=True,
+)
+```
+
+## Bash guidelines
+
+> [!WARNING]
+> Bash is **strongly discouraged** for nontrivial usage in .yml GitHub Actions
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:422]: maybe add a comment that it also creates problem for the different platform support (win vs linux)

File: docs/development/style_guide.md [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+- `set -o pipefail` Uses the return code of a failing command as the pipeline
+  return code
+- `set -x` Prints commands to the terminal (useful for debugging and CI logging)
+
+See https://gist.github.com/mohanpedala/1e2ff5661761d3abd0385e8223e16425 for
+an explanation of what each of these options does.
+
+## GitHub Actions guidelines
+
+### Pin action versions to commit SHAs
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: `+` add a comment what to what the sha refers to :)

File: docs/development/style_guide.md [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+
+### Prefer Python scripts over inline Bash
+
+Where possible, put workflow logic in Python scripts.
+
+Python scripts have several benefits:
+
+- **Testable:** Can be tested locally and with unit tests
+- **Debuggable:** Easier to debug with standard Python tools
+- **Portable:** Works consistently across platforms (Linux/Windows/macOS)
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md]: do we support anything for macos? otherwise i would remove it

File: docs/development/style_guide.md [hunk 15]
```diff
@@ -0,0 +1,583 @@
...
+```yaml
+- name: Process artifacts
+  run: |
+    python build_tools/process_artifacts.py \
+      --families "${{ inputs.amdgpu_families }}" \
+      --artifact-dir artifacts \
+      --install-dir install
+```
+
+### Use safe defaults for inputs
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:552]: this should be the case for all workflows + cli scripts?

File: docs/development/style_guide.md [hunk 16]
```diff
@@ -0,0 +1,583 @@
...
+
+Workflow inputs should have safe default values that work in common scenarios.
+
+Good defaults should:
+
+- **Work without configuration:** Safe defaults that won't trigger production changes
```
Comment (HereThereBeDragons) [inline docs/development/style_guide.md:558]: **ALWAYS use `dev` as default!!!**

File: docs/development/style_guide.md (hunk 8 above)
Comment (ScottTodd) [reply docs/development/style_guide.md]: > unless you come from C you dont do manual parsing anyway

If only... I've seen this code very recently: https://github.com/ROCm/rocm-libraries/pull/2411/files#r2519049187

Image

I think it's worth keeping the example so people can pattern match code "oh yeah, I saw that somewhere, let me go fix it".

File: docs/development/style_guide.md (hunk 11 above)
Comment (ScottTodd) [reply docs/development/style_guide.md:327]: I agree, though I think we should split this into a few separate style rules. The `__main__` guard is one that is especially helpful for library development. See how long it takes to run `import torch` for example - that runs a lot of code and imports a lot of other files eagerly (vs lazily).

The point about `def run(args)` is about designing for modularity and unit testing. It is easier to write clear unit tests for functions that take 2 string arguments than it is for functions that take `args` (who knows what's in there and what is or isn't relevant to that particular test case).

File: docs/development/style_guide.md (hunk 15 above)
Comment (ScottTodd) [reply docs/development/style_guide.md:552]: Sure. I wanted this to be specific to actions but we should have a similar rule/guideline for scripts.

I added a note about "stricter security boundaries" hinting at cases where we won't even allow overriding. We can refine that once we have more workflows moved over to https://github.com/ROCm/rockrel.

File: docs/development/style_guide.md (hunk 10 above)
Comment (ScottTodd) [reply docs/development/style_guide.md:198]: I added a boolean argument with an example of using `parser.error` too

File: docs/development/style_guide.md (hunk 7 above)
Comment (ScottTodd) [reply docs/development/style_guide.md]: I adjusted the examples a bit but I didn't take this suggestion [yet]. I see your point though...

**Review by HereThereBeDragons — APPROVED** [review]
the `pre-commit` part is still in `CONTRIBUTING.md`. We can keep it at both places or not.

And i would still add that the style guide is more than a style guide

otherwise lgtm :)

**Comment by ScottTodd** [comment]
> the `pre-commit` part is still in `CONTRIBUTING.md`. We can keep it at both places or not.

I don't mind the duplication.

> And i would still add that the style guide is more than a style guide

I added a note at the top of the guide encouraging contributions/discussion. I think calling it a "style guide" isn't _too_ much of a stretch, though some of the points are closer to recommendations like those on https://testing.googleblog.com/ than they are style rules. I am also considering splitting into one page per language/tool.
